zvariant/src/fd.rs

```diff
@@ -1,155 +1,134 @@
 
-fd_impl!(Fd);
+assert_impl_all!(Fd: Send, Sync, Unpin);
 
-impl Serialize for Fd {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        serializer.serialize_i32(self.0)
+impl Basic for Fd {
+    const SIGNATURE_CHAR: char = 'h';
+    const SIGNATURE_STR: &'static str = "h";
+
+    fn alignment(format: EncodingFormat) -> usize {
+        u32::alignment(format)
     }
 }
 
-impl<'de> Deserialize<'de> for Fd {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        Ok(Fd(i32::deserialize(deserializer)?))
+impl Type for Fd {
+    fn signature() -> Signature<'static> {
+        Signature::from_static_str_unchecked(Self::SIGNATURE_STR)
     }
 }
 
-impl From<io::RawFd> for Fd {
-    fn from(value: io::RawFd) -> Self {
-        Self(value)
+impl<T: Into<OwnedFd>> From<T> for Fd {
+    fn from(fd: T) -> Self {
+        Self(fd.into())
     }
 }
 
-impl<T> From<&T> for Fd
-where
-    T: io::AsRawFd,
-{
-    fn from(t: &T) -> Self {
-        Self(t.as_raw_fd())
+impl FromRawFd for Fd {
+    unsafe fn from_raw_fd(fd: RawFd) -> Self {
+        Self(OwnedFd::from_raw_fd(fd))
     }
 }
 
-impl io::AsRawFd for Fd {
-    fn as_raw_fd(&self) -> io::RawFd {
-        self.0
+impl AsRawFd for Fd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.0.as_raw_fd()
     }
 }
 
-impl std::fmt::Display for Fd {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        self.0.fmt(f)
+impl AsFd for Fd {
+    fn as_fd(&self) -> BorrowedFd<'_> {
+        self.0.as_fd()
     }
 }
 
-/// An owned [`RawFd`](https://doc.rust-lang.org/std/os/unix/io/type.RawFd.html) wrapper.
-///
-/// See also [`Fd`]. This type owns the file and will close it on drop. On deserialize, it will
-/// duplicate the file descriptor.
-#[derive(Debug, PartialEq, Eq, Hash)]
-pub struct OwnedFd {
-    inner: io::RawFd,
+#[derive(Debug)]
+pub struct FdWrapper(Arc<Fd>);
+
+assert_impl_all!(FdWrapper: Send, Sync, Unpin);
+
+impl FdWrapper {
+    pub(crate) fn new(fd: Fd) -> Self {
+        Self(Arc::new(fd))
+    }
+
+    pub fn into_inner(self) -> Arc<Fd> {
+        self.0
+    }
 }
 
-impl Drop for OwnedFd {
-    fn drop(&mut self) {
-        unsafe {
-            libc::close(self.inner);
-        }
+impl Clone for FdWrapper {
+    fn clone(&self) -> Self {
+        Self(Arc::clone(&self.0))
     }
 }
 
-fd_impl!(OwnedFd);
+impl PartialEq for FdWrapper {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.as_raw_fd() == other.0.as_raw_fd()
+    }
+}
 
-impl Serialize for OwnedFd {
+impl Serialize for FdWrapper {
     fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
     where
         S: Serializer,
     {
-        serializer.serialize_i32(self.inner)
+        let dupped = self.0 .0.try_clone().map_err(Error::custom)?;
+        let raw_fd = dupped.into_raw_fd();
+        serializer.serialize_i32(raw_fd)
```

Comment:
Looks like this code leaks fds?

Also, I'm not familiar with the way zvariant handles fds, but what is the point of serializing and fd as i32? Seems unnecessary.

Reply:
> Looks like this code leaks fds?
> 
> Also, I'm not familiar with the way zvariant handles fds, but what is the point of serializing and fd as i32? Seems unnecessary.

To send fds across over D-Bus. It's part of the spec.

Reply:
Ah, I dug into Serializer implementation a little more, it knows if a given i32 is an fd by looking at the signature, nvm :facepalm:

Reply:
Anyway, it's unfortunate that it is not possible to pass `Arc<Fd>` directly to the serializer without duplicating the fd.

Reply:
I mean, it is possible to clone the `Arc`, call `into_raw()`, and pass the pointer as `u64`...

Reply:
> I mean, it is possible to clone the `Arc`, call `into_raw()`, and pass the pointer as `u64`...

Yeah. I think we had it all figured out in the end after many discussions at GUADEC.